docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,410 @@
+{
```

Comment:
<p>I spent some time on refactoring this code and came up with this alternative:</p><p> </p><pre class="ql-syntax" spellcheck="false">import numpy.testing as npt


def _select_candidates(subseqs, min_dist):
&nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; n, m = subseqs.shape
&nbsp; &nbsp; cand_idx = np.zeros(n, dtype=bool)
&nbsp; &nbsp; excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
&nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; r = m * (1.0 - ((min_dist ** 2.0) / (2.0 * m)))&nbsp;
&nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; for i in range(0, n):
&nbsp; &nbsp; &nbsp; &nbsp; add_cand_i = True
&nbsp; &nbsp; &nbsp; &nbsp; if i &gt;= excl_zone:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; non_trivial_cand_idx = np.flatnonzero(cand_idx[:i-excl_zone+1])
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; R = np.matmul(subseqs[non_trivial_cand_idx], subseqs[i])
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; mask = np.flatnonzero(R &gt; r)
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; cand_idx[non_trivial_cand_idx[mask]] = False

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; if len(mask):
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; add_cand_i = False

&nbsp; &nbsp; &nbsp; &nbsp; if add_cand_i:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; cand_idx[i] = True
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; return np.flatnonzero(cand_idx)


_cand_index = _select_candidates(subsequences, min_dist)
npt.assert_almost_equal(_cand_index, cand_index)
</pre><p> </p><p>I'm probably biased but I find the logic here to be more obvious and much easier to follow. I also renamed certain variables accordingly. The last two lines tests and makes sure that the output between my implementation and your implementation are the same (i.e., given the same input, they produce the same output). Though, I have not tested it on other data sets (please do this!).</p><p> </p><p>What do you think?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Reply:
I think your approach is better as there  is no need to recreate the `cand_idx` again and again. So, it should be faster. I will test it.

Reply:
<p>In my limited timing test for this small dataset, my implementation was ever so slightly faster but, honestly, negligible. </p>

Reply:
Thanks for letting me know. I am not sure if the difference in the running time stays negligible in different cases. I will go with your approach anyway. 

Just out of curiosity, if the running time is almost the same (and not an issue), may I know your reason behind using `cand_idx = np.zeros(n, dtype=bool)` rather than re-creating `cand_idx` in each iteration? Because, as I mentioned in the notebook, the number of candidates are about 1/10 of total number of subsequences. Will your approach be okay for VERY VERY LONG TIME SERIES? because, we are wasting some memory. Right?

Reply:
My approach should be fine. I wouldn't expect any input time series to be longer than say 100,000,000 data points and NumPy can create this boolean array instantly and it only needs to do it once. Efficiency is always a balance between memory and time. In this case, we are trying to optimize for time since the memory usage for `cand_idx` only grows linearly with the size of the data. The benefit of the current implementation, i think, is the readability. In my brain, it just makes sense to set "good" indices to `True` and "bad" indices to `False` and it doesn't take much mental effort to think about what it means. This should make it much easier to maintain IMHO

I would be more concerns with storage of z-norm subsequences taking up a ton of memory

Reply:
Although, keep in mind that arrays like `non_trivial_cand_idx`, `R`, and `mask` are being created over-and-over again. This can still be bad for larger time series so we should test this. Of course, we should always compare this against the results and the timing to `stumpy.stump` for, say, a time series of length `n=50000` to `n=100000`.

One more thing before I forget. I see that the MERLIN paper uses a much larger exclusion zone for discords but I think it is more important to be consistent with `stumpy.stump` discord output so we'll have to remember to add a note to the docstring about this difference.

Reply:
> Although, keep in mind that arrays like non_trivial_cand_idx, R, and mask are being created over-and-over again. This can still be bad for larger time series so we should test this. Of course, we should always compare this against the results and the timing to stumpy.stump for, say, a time series of length n=50000 to n=100000.

Correct. That might be a burden. I can also do an primary test on my end as I already implemented MERLIN on another notebook. 


> I see that the MERLIN paper uses a much larger exclusion zone for discords but I think it is more important to be consistent with stumpy.stump discord output so we'll have to remember to add a note to the docstring about this difference.


Yes. We first considered the exclusion considered in the original paper. Then, we decided to change it to stumpy' default to make sure the output of MERLIN is consistent with output of stumpy. Will put a note in docstring.

---

Also, I noticed you already hit "resolved" for one of our conversation. So, should I wait till you hit resolved for all of them and then I start pushing the updated notebook to the repo? 
Or, Should I resolve them myself if I think they are already discussed and wait, say, for a day, and see if you have anything more to say?

Reply:
Hmmm, how about this (and please discuss/debate as you have as much input as I do), since you are the one actually resolving the comments then,

1. Please debate, discuss, seek clarification, share potential implementation details, etc
2. Once you feel like you have what you need from step 1 to resolve the comment then please leave a comment to the extent of "I will handle this in the next commit" or "I think I have enough to resolve this commit"
3. And hit "Resolved conversation" after you've completed things for that comment
4. Then, when all of the comments for this commit series are ALL resolved then please leave me a note (in the global PR, not in the notebook) along the lines of "All comments are resolved now" and wait a day for me to give the green light for you to push the next set of commits or maybe I might respond asking for more time or more details on the already resolved comments

How does that sound?

Reply:
That is very well structured. Sounds great to me!!

Reply:
Btw, I find it funny that I basically came to the same implementation as the paper. While I tried to refactor your code without looking at the paper, it is good to know that their implementation is likely as simple as possible! There is actually a good opportunity to parallelize:

```
non_trivial_cand_idx = np.flatnonzero(cand_idx[:i-excl_zone+1])
R = np.matmul(subseqs[non_trivial_cand_idx], subseqs[i])
mask = np.flatnonzero(R > r)
cand_idx[non_trivial_cand_idx[mask]] = False
```

using `numba` but we can look at this in a future iteration

Reply:
> it is good to know that their implementation is likely as simple as possible!

I should try to follow their implementation more closely. I thought I was making it simpler. Apparently I was decreasing its readability. 


> using `numba` but we can look at this in a future iteration

I can add a small note to have that in mind. 

Also, in `fastMERLIN`, the author changes the current `for-loop` (on subsequences) to `for-loop` on blocks of subsequence. If, in the future, we decide to go for `fastMERLIN`, we can parallelize the `within-comparison` of each block as well. 

---

I also noticed that you used underscore `_` as a prefix to the name of variable and function ( (I think) because these functions are being used internally and not by the end user). I will keep that in mind.

---

I can now start updating the function. I will address the discussed issues here in the next commit.

Reply:
<p>I thought about it a little more. Instead of <code>non_trivial_cand_idx</code> can you please call it <code>left_non_trivial_cand_idx</code>?  I know that it is a bit long but it really makes it obvious to the reader what that variable is capturing precisely </p>

Reply:
Sure. I can use `left_non_trivial_cand_idx` for this function. 

In the next function where the candidates are being pruned (by being compared to their LEFT neighbors), I cannot use this name as it is misleading, and I should better not to use `right_non_trivial_cand_idx` as it may wrongfully deliver a message that says we are getting new candidates. Therefore, I will just use `non_trivial_cand_idx` for that second function. (Then, you can do a review and see if that makes sense or we should change the name of candidates in that function as well)

Reply:
Also, if I understand correctly, `cand_idx[:i-excl_zone+1]` should be changed to `cand_idx[: i-excl_zone]` as the ` i-excl_zone` itself is in the exclusion area for `subseqs[i]`

Reply:
<blockquote>Then, you can do a review and see if that makes sense or we should change the name of candidates in that function as well</blockquote><p>Sounds good. I'll look out for it</p><p> </p><blockquote>Also, if I understand correctly,&nbsp;<code>cand_idx[:i-excl_zone+1]</code>&nbsp;should be changed to&nbsp;<code>cand_idx[: i-excl_zone]</code>&nbsp;as the&nbsp;<code> i-excl_zone</code>&nbsp;itself is in the exclusion area for&nbsp;<code>subseqs[i]</code>. I changed this and the&nbsp;<code>if</code>&nbsp;condition in the code.</blockquote><p>Yes! Good catch. The goal of my implementation was to ensure that it produced the same output as your original version. So, if there is an error there then it will persist in my code as well. It should be <code>cand_idx[:i-excl_zone]</code>. since this should exclude <code>i-excl_zone</code>  itself.</p>

Reply:
> The goal of my implementation was to ensure that it produced the same output as your original version.

Exactly. You wanted to focus on the "refactoring" thing and just clean the same code. All good!

Reply:
<p>I've been thinking about something, is there a reason why we don't update <code>min_dist</code> or <code>r</code>    when we find a new candidate that has a dot product that is smaller than <code>r</code> ? In other words:</p><p> </p><pre class="ql-syntax" spellcheck="false">if len(mask):
    add_cand_i = False

if len(left_non_trivial_cand_idx):
    r = min(R.min(), r)
</pre><p> </p><p>I don't know if this will produce incorrect results but it makes things almost 10x faster! Then, at the end, we should return <code>np.flatnonzero(cand_idx), new_min_dist</code>  (where we convert <code>r</code>  back to <code>min_dist</code>  and then maybe we use this <code>new_min_dist</code>  for future calculations.</p>

Reply:
That seems to be a very brilliant idea!!! Please let me think about it. I will try to use it in my fully-implemented MERLIN (in another notebook) and see how things go!  

Happy New Year to you too!!!

Reply:
<p>Weird, I tried this on a time series <code>T = np.random.rand(100000)</code> and this was twice as slow :(</p><p> </p><p>Please think about it :)</p>

Reply:
> Weird, I tried this on a time series `T = np.random.rand(100000)` and this was twice as slow :(
> 
> But I think you understand my rationale so please think about it :)

So, I thought about how it can affect the behavior of the algorithm. This is what I think:
Let us assume the distance of discord to its NN is `d`. Now, let us review four scenarios (I also included one infeasible scenario):
* Scenario (0-infeasible): if `min_dist` is very large! (e.g. `2*sqrt(m) + 1e-6`). In this case, NO candidates are going to be returned as all distances will be less than `min_dist` (recall that ` 2*sqrt(m)` is the maximum possible z-normalized Euclidean distance)
*  Scenario (1): if `min_dist` is the highest possible value (i.e. `2*sqrt(m)`). In this case, we should be lucky to "quickly" find the discord. That can happen if `d = 2*sqrt(m)`.
* Scenario (2): if `min_dist = d + 1e-6` (i.e. slightly bigger than `d`): STILL we canNOT find any discord.
* Scenario (3): if `min_dist = d - 1e-6` (i.e. slightly smaller than `d`): we get a few candidates and we can quickly find discord
* Scenario (4): if `min_dist` is very small! (e.g. 0). In this case, all subsequences are going to be selected as candidates! So, we CAN find the discord BUT it takes a lot of time as we have a LOT of candidates.

Now, suppose we are at scenario (3). If we want to update 'min_dist' in the middle of the algorithm (i.e. making it larger to prune more), we cannot know if we are getting closer to Scenario 3 (from below) or we are getting closer to scenario (3) from above or even maybe getting into scenario (2), (1), or somewhere in between!

To better understand this, let us assume we have just two so-far-selected candidates (say A and B) and we are comparing them against the subsequent s = `T[i: i+m]`. Now, what if s is really far from A and B? we cannot say A/B/s is discord because each of them may have a very close NN (just think of them as being in two different clusters being far from each other). So, their pair-wise distance from each other (i.e. dist(A,s) and/or dist(B,s)) can be `d + something`!!! and it may take us to Scenario (2) or (1)!

Now, on the other hand, if we just lower the threshold little by little (as  proposed in the paper and without updating it in the middle of process), we know that we are in a "stable" direction to find SOME candidates at some point in the process!

---

Please let me know what you think. Not sure how clearly I explained my thoughts.

Reply:
I just tried it in the `_select_candidates` function. I realized the output doesn't contain the index of discord. It just contains the false positive ones that are going to be pruned in the next step (I checked). Therefore, it seems it is better to not update `r` (or `min_dist`).